File: ast.cpp
#include <cctype>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

enum class TokenType {
    Number,
    Operator,
    LeftParen,
    RightParen
};

struct Token {
    TokenType type;
    union {
        int number;
        char op;
    };

    Token(int num)
        : type(TokenType::Number), number(num) {}
    Token(char op)
        : type(TokenType::Operator), op(op) {}
    Token(TokenType t)
        : type(t) {
        if (t != TokenType::LeftParen && t != TokenType::RightParen) {
            throw std::invalid_argument(std::string("invalid delimiter kind"));
        }
    }
};

class AST {
  public:
    virtual ~AST() = default;
    virtual int evaluate() const = 0;
};

class NumberAST : public AST {
    int value;

  public:
    NumberAST(int v)
        : value(v) {}
    int evaluate() const override { return value; }
};

class BinaryOpAST : public AST {
    char op;
    std::unique_ptr<AST> left;
    std::unique_ptr<AST> right;

  public:
    BinaryOpAST(char op_, std::unique_ptr<AST> l, std::unique_ptr<AST> r)
        : op(op_), left(std::move(l)), right(std::move(r)) {}

    int evaluate() const override {
        int l_val = left->evaluate();
        int r_val = right->evaluate();
        switch (op) {
        case '+': return l_val + r_val;
        case '-': return l_val - r_val;
        case '*': return l_val * r_val;
        case '/': return l_val / r_val;
        default: throw std::invalid_argument(std::string("invalid token `") + op + "`");
        }
    }
};

std::vector<Token> lex(const std::string &expr) {
    std::vector<Token> tokens;
    size_t i = 0;
    while (i < expr.size()) {
        if (std::isdigit(expr[i])) {
            int num = 0;
            while (i < expr.size() && std::isdigit(expr[i])) {
                num = num * 10 + (expr[i] - '0');
                i++;
            }
            tokens.emplace_back(num);
        } else if (expr[i] == '+' || expr[i] == '-' || expr[i] == '*' || expr[i] == '/') {
            tokens.emplace_back(expr[i]);
            i++;
        } else if (expr[i] == '(') {
            tokens.emplace_back(TokenType::LeftParen);
            i++;
        } else if (expr[i] == ')') {
            tokens.emplace_back(TokenType::RightParen);
            i++;
        } else if (std::isspace(expr[i])) {
            i++;
        } else {
            throw std::runtime_error(std::string("unknown character: ") + expr[i]);
        }
    }
    return tokens;
}

int precedence(char op) {
    switch (op) {
    case '+':
    case '-': return 1;
    case '*':
    case '/': return 2;
    default: return 0;
    }
}

std::unique_ptr<AST> parse(const std::vector<Token> &tokens) {
    std::vector<std::unique_ptr<AST>> output_stack;
    std::vector<Token> operator_stack;

    for (const Token &token : tokens) {
        switch (token.type) {
        case TokenType::Number: {
            output_stack.push_back(std::make_unique<NumberAST>(token.number));
            break;
        }
        case TokenType::Operator: {
            while (!operator_stack.empty() &&
                   operator_stack.back().type == TokenType::Operator &&
                   precedence(operator_stack.back().op) >= precedence(token.op)) {
                char op = operator_stack.back().op;
                operator_stack.pop_back();

                if (output_stack.size() < 2) {
                    throw std::runtime_error("invalid expression");
                }

                auto right = std::move(output_stack.back());
                output_stack.pop_back();
                auto left = std::move(output_stack.back());
                output_stack.pop_back();

                output_stack.push_back(
                    std::make_unique<BinaryOpAST>(op, std::move(left), std::move(right)));
            }
            operator_stack.push_back(token);
            break;
        }
        case TokenType::LeftParen: {
            operator_stack.push_back(token);
            break;
        }
        case TokenType::RightParen: {
            while (!operator_stack.empty() && operator_stack.back().type != TokenType::LeftParen) {
                if (operator_stack.back().type != TokenType::Operator) {
                    throw std::runtime_error("invalid operator stack state");
                }
                char op = operator_stack.back().op;
                operator_stack.pop_back();

                if (output_stack.size() < 2) {
                    throw std::runtime_error("invalid expression");
                }

                auto right = std::move(output_stack.back());
                output_stack.pop_back();
                auto left = std::move(output_stack.back());
                output_stack.pop_back();

                output_stack.push_back(
                    std::make_unique<BinaryOpAST>(op, std::move(left), std::move(right)));
            }
            if (operator_stack.empty()) {
                throw std::runtime_error("mismatched parentheses");
            }
            operator_stack.pop_back();
            break;
        }
        }
    }

    while (!operator_stack.empty()) {
        if (operator_stack.back().type != TokenType::Operator) {
            throw std::runtime_error("mismatched parentheses");
        }
        char op = operator_stack.back().op;
        operator_stack.pop_back();

        if (output_stack.size() < 2) {
            throw std::runtime_error("invalid expression");
        }

        auto right = std::move(output_stack.back());
        output_stack.pop_back();
        auto left = std::move(output_stack.back());
        output_stack.pop_back();

        output_stack.push_back(
            std::make_unique<BinaryOpAST>(op, std::move(left), std::move(right)));
    }

    if (output_stack.size() != 1) {
        throw std::runtime_error("invalid expression");
    }

    return std::move(output_stack.back());
}

int main() {
    std::string expr;
    std::cout << "in> ";
    std::getline(std::cin, expr);
    try {
        auto tokens = lex(expr);
        auto ast = parse(tokens);
        int result = ast->evaluate();
        std::cout << "out> " << result << std::endl;
    } catch (const std::exception &e) {
        std::cerr << "err> " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: ast.rs
#[derive(Debug, Clone, PartialEq)]
enum Token {
    Number(i32),
    Operator(char),
    LeftParen,
    RightParen,
}

#[derive(Debug)]
enum AST {
    Number(i32),
    BinaryOp {
        op: char,
        left: Box<AST>,
        right: Box<AST>,
    },
}

fn lex(expr: &str) -> Result<Vec<Token>, String> {
    let mut tokens = Vec::new();
    let mut chars = expr.chars().peekable();

    while let Some(&ch) = chars.peek() {
        match ch {
            '0'..='9' => {
                let mut num = 0;
                while let Some(digit @ '0'..='9') = chars.peek() {
                    num = num * 10 + digit.to_digit(10).unwrap() as i32;
                    chars.next();
                }
                tokens.push(Token::Number(num));
            }
            '+' | '-' | '*' | '/' => {
                tokens.push(Token::Operator(ch));
                chars.next();
            }
            '(' => {
                tokens.push(Token::LeftParen);
                chars.next();
            }
            ')' => {
                tokens.push(Token::RightParen);
                chars.next();
            }
            ' ' => {
                chars.next();
            }
            _ => return Err(format!("unknown character: {}", ch)),
        }
    }

    Ok(tokens)
}
fn parse(tokens: &[Token]) -> AST {
    let mut output_stack: Vec<AST> = Vec::new();
    let mut operator_stack: Vec<Token> = Vec::new();

    let precedence = |op: char| match op {
        '+' | '-' => 1,
        '*' | '/' => 2,
        _ => 0,
    };

    for token in tokens {
        match token {
            Token::Number(n) => output_stack.push(AST::Number(*n)),
            Token::Operator(op) => {
                while let Some(Token::Operator(top_op)) = operator_stack.last() {
                    if precedence(*top_op) >= precedence(*op) {
                        let right = output_stack.pop().unwrap();
                        let left = output_stack.pop().unwrap();
                        output_stack.push(AST::BinaryOp {
                            op: *top_op,
                            left: Box::new(left),
                            right: Box::new(right),
                        });
                        operator_stack.pop();
                    } else {
                        break;
                    }
                }
                operator_stack.push(Token::Operator(*op));
            }
            Token::LeftParen => operator_stack.push(Token::LeftParen),
            Token::RightParen => {
                while let Some(top) = operator_stack.pop() {
                    if let Token::LeftParen = top {
                        break;
                    }
                    let right = output_stack.pop().unwrap();
                    let left = output_stack.pop().unwrap();
                    if let Token::Operator(op) = top {
                        output_stack.push(AST::BinaryOp {
                            op,
                            left: Box::new(left),
                            right: Box::new(right),
                        });
                    }
                }
            }
        }
    }

    while let Some(top) = operator_stack.pop() {
        let right = output_stack.pop().unwrap();
        let left = output_stack.pop().unwrap();
        if let Token::Operator(op) = top {
            output_stack.push(AST::BinaryOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            });
        }
    }

    output_stack.pop().unwrap()
}
fn evaluate(ast: &AST) -> Result<i32, String> {
    match ast {
        AST::Number(n) => Ok(*n),
        AST::BinaryOp { op, left, right } => {
            let left_val = evaluate(left).unwrap_or_else(|e| {
                println!("err> {e}");
                std::process::exit(1)
            });
            let right_val = evaluate(right).unwrap_or_else(|e| {
                println!("err> {e}");
                std::process::exit(1)
            });
            match op {
                '+' => Ok(left_val + right_val),
                '-' => Ok(left_val - right_val),
                '*' => Ok(left_val * right_val),
                '/' => Ok(left_val / right_val),
                _ => Err(String::from("unknown operator")),
            }
        }
    }
}
fn main() {
    let mut expression = String::new();
    print!("in~> ");
    use std::io::Write;
    std::io::stdout().flush().unwrap();
    std::io::stdin().read_line(&mut expression).unwrap();

    let tokens = lex(&expression[..expression.len() - 1]).unwrap_or_else(|e| {
        println!("err> {e}");
        std::process::exit(1)
    });
    let ast = parse(&tokens);
    let result = evaluate(&ast).unwrap_or_else(|e| {
        println!("err> {e}");
        std::process::exit(1)
    });
    println!("out> {}", result);
}
